[Feature blogpost][1.22] CSI Windows Support with CSI Proxy reaches GA (#29105)
kubernetes/website: content/en/blog/_posts/2021-07-27-csi-windows-support-with-csi-proxy-reaches-ga.md

---
layout: blog
title: 'Kubernetes 1.22: CSI Windows Support with CSI Proxy reaches GA'
date: 2021-07-27
slug: csi-windows-support-with-csi-proxy-reaches-ga
---

**Authors:** Mauricio Poppe (Google), Jing Xu (Google), and Deep Debroy (Apple)

> *The GA version of CSI Proxy for Windows is being released in Kubernetes 1.22.  CSI Proxy enables CSI Drivers running on Windows nodes to perform privileged storage operations.*

Reviewer: If this goes out post-release, the text should be:
```suggestion
*The stable version of CSI Proxy for Windows has been released alongside Kubernetes 1.22.  CSI Proxy enables CSI Drivers running on Windows nodes to perform privileged storage operations.*
```

Author: ack

- The controller plugin generally does not need direct access to the host and can perform all its operations through the Kubernetes API and external control plane services.

- The node plugin, however, requires direct access to the host for making block devices and/or file systems available to the Kubernetes kubelet. Due to the missing capability of running privileged operations from containers on Windows nodes [CSI Proxy was introduced as alpha in Kubernetes 1.18](https://kubernetes.io/blog/2020/04/03/kubernetes-1-18-feature-windows-csi-support-alpha/) as a way to enable containers to perform privileged storage operations. This enables containerized CSI Drivers to run on Windows nodes.

## What's CSI Proxy and how do CSI drivers interact with it?

When a workload that uses persistent volumes is scheduled, it'll go through a sequence of steps defined in the [CSI Spec](https://github.com/container-storage-interface/spec/blob/master/spec.md). First, the workload will be scheduled to run on a node. Then the controller component of a CSI Driver will attach the persistent volume to the node. Finally the node component of a CSI Driver will mount the persistent volume on the node.

> The node component of a CSI Driver needs to run on Windows nodes to support Windows workloads. Various privileged operations like scanning of disk devices, mounting of file systems, etc. cannot be done from a containerized application running on Windows nodes yet (Windows Host Process is available in kubernetes 1.22 as alpha). However, we can perform these operations through a binary (CSI Proxy) that's pre-installed on the Window nodes. CSI Proxy has a client-server architecture and allows CSI drivers to issue privileged storage operations through a gRPC interface exposed over named pipes created during the startup of CSI Proxy.

Author: Changed to:

> [Windows HostProcess containers](https://github.com/kubernetes/enhancements/issues/1981) introduced in Kubernetes 1.22 as alpha enable functionalities that require host access like the operations mentioned before

![CSI Proxy Architecture](/images/blog/2021-07-27-csi-windows-support-with-csi-proxy-reaches-ga/csi-proxy.png)

## CSI Proxy reaches GA

> [Since the introduction of the CSI Proxy KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1122-windows-csi-support) storage vendors integrated CSI Proxy into their CSI Drivers and provided feedback about use cases where we needed new APIs as well as bug reports, bug fixes and documentation updates.

Reviewer: ```suggestion
Since the introduction of the [CSI Proxy KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1122-windows-csi-support), storage vendors integrated CSI Proxy into their CSI Drivers and provided feedback. We learned about use cases where we needed new APIs, as well as getting bug reports, bug fixes and documentation updates.
```

Author: ack

CSI Proxy is compatible with all the previous v1betaX releases. The `csi-proxy.exe` binary deployed on Windows nodes still serves requests from v1betaX clients thanks to the autogenerated conversion layer that transforms any versioned client request to a version-agnostic request that the server can process. We added several [integration tests](https://github.com/kubernetes-csi/csi-proxy/tree/v1.0.0/integrationtests) for all the API versions of the API groups that are graduating to v1 to ensure that CSI Proxy is backwards compatible.

> We've also considered the scenario of a version drift between CSI Proxy and the CSI Drivers that interact with it and provided a way for CSI Drivers to perform a smooth upgrade to v1, GCE PD CSI Driver can recognize which version of the CSI Proxy binary is running and is able to handle multiple versions of the CSI Proxy binary deployed on the node.

Reviewer 2: Use word like "For example" before GCE PD CSI Driver part? Also put a link for it.

Reviewer 2: and a little bit detail like "pd csi driver will try to access V1 API first, if it is not available, it will try to communicate with beta API"?

Author: Changed some things in this paragraph.

> We've also considered the scenario of a version drift between CSI Proxy and the CSI Drivers that interact with it and provided a way for CSI Drivers to handle multiple versions of CSI Proxy for a smooth upgrade to v1. [For example](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/738), the GCE PD CSI Driver will use the CSI Proxy client libraries to connect to the v1 API first, if the connection is unsuccessful because the v1 version is not installed it'll connect to the v1beta API.

Reviewer 3: ```suggestion
CSI Proxy supports version drift between CSI Proxy version and the CSI Drivers that interact with it, ensuring a smooth upgrade to v1. [For example](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/738), the GCE PD CSI Driver uses the CSI Proxy client libraries to connect to the v1 API first, if the connection is unsuccessful because the v1 version is not installed it'll fallback to v1beta.
```

> CSI Proxy v1 is already being used by many CSI Drivers, such as the [AWS EBS CSI Driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/966), [Azure Disk CSI Driver](https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/919), [GCE PD CSI Driver](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/738) and [SMB CSI Driver](https://github.com/kubernetes-csi/csi-driver-smb/pull/319).

Reviewer 3: ```suggestion
CSI Proxy v1 is already being used by many CSI Drivers, including the [AWS EBS CSI Driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/966), [Azure Disk CSI Driver](https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/919), [GCE PD CSI Driver](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/738), and [SMB CSI Driver](https://github.com/kubernetes-csi/csi-driver-smb/pull/319).
```

I'm a big fan of the oxford comma 😄

## Future plans

We're very excited for the future of CSI Proxy. With the upcoming [support for privileged Windows containers](https://github.com/kubernetes/enhancements/issues/1981), we plan to use CSI Proxy as a library in CSI Drivers in addition to the current client/server design. This will allow us to iterate faster on new features because  the `csi-proxy.exe` binary will no longer be needed.

> API support for the iSCSI protocol is in the alpha stage, we plan to make additional enhancements before graduating it to v1.

Reviewer 2: also system API https://github.com/kubernetes-csi/csi-proxy/tree/master/client/api/system/v1alpha1

Author: Added a link for the system API

> We're very excited for the future of CSI Proxy. With the upcoming [support for privileged Windows containers](https://github.com/kubernetes/enhancements/issues/1981), we plan to use CSI Proxy as a library in CSI Drivers in addition to the current client/server design. This will allow us to iterate faster on new features because  the `csi-proxy.exe` binary will no longer be needed.

Reviewer 3: ```suggestion
We're very excited for the future of CSI Proxy. With the upcoming [support for privileged Windows containers](https://github.com/kubernetes/enhancements/issues/1981), we are considering converting the CSI Proxy in to a library consumed by CSI Drivers in addition to the current client/server design. This will allow us to iterate faster on new features because  the `csi-proxy.exe` binary will no longer be needed.
```

"plan" seems a little too strong :)

> Support for the [System API](https://github.com/kubernetes-csi/csi-proxy/tree/v1.0.0/client/api/system/v1alpha1) and the [iSCSI API](https://github.com/kubernetes-csi/csi-proxy/tree/v1.0.0/client/api/iscsi/v1alpha2) is in the alpha stage, we plan to make additional enhancements before graduating them to v1.

Reviewer 3: This seems out of place, either add more context about what it is and how it relates to CSI Proxy or just remove it.

Reviewer 2: maybe move this part in line 37/38 along with APIs reach V1

Author: Changed to:

> Windows system APIs to get additional information from the Windows nodes and support to mount iSCSI targets in Windows nodes are available as alpha APIs in the [System API](https://github.com/kubernetes-csi/csi-proxy/tree/v1.0.0/client/api/system/v1alpha1) and the [iSCSI API](https://github.com/kubernetes-csi/csi-proxy/tree/v1.0.0/client/api/iscsi/v1alpha2), these APIs will continue to be improved before we graduate them to v1.

Also moved it to be along the APIs that are reaching v1.
